src/java.base/macosx/classes/sun/nio/ch/KQueue.java

```diff
@@ -47,74 +59,112 @@ private KQueue() { }
 
     // filters
-    static final int EVFILT_READ  = -1;
-    static final int EVFILT_WRITE = -2;
+    static final int EVFILT_READ  = kqueue_h.EVFILT_READ();
+    static final int EVFILT_WRITE = kqueue_h.EVFILT_WRITE();
 
     // flags
-    static final int EV_ADD     = 0x0001;
-    static final int EV_DELETE  = 0x0002;
-    static final int EV_ONESHOT = 0x0010;
-    static final int EV_CLEAR   = 0x0020;
+    static final int EV_ADD     = kqueue_h.EV_ADD();
+    static final int EV_DELETE  = kqueue_h.EV_DELETE();
+    static final int EV_ONESHOT = kqueue_h.EV_ONESHOT();
+    static final int EV_CLEAR   = kqueue_h.EV_CLEAR();
 
     /**
      * Allocates a poll array to handle up to {@code count} events.
      */
-    static long allocatePollArray(int count) {
-        return unsafe.allocateMemory(count * SIZEOF_KQUEUEEVENT);
+    static MemorySegment allocatePollArray(int count) {
+        return kevent.allocateArray(count, FFMUtils.SEGMENT_ALLOCATOR);
     }
 
     /**
      * Free a poll array
      */
-    static void freePollArray(long address) {
-        unsafe.freeMemory(address);
+    static void freePollArray(MemorySegment memorySegment){
+        FFMUtils.free(memorySegment);
     }
 
     /**
      * Returns kevent[i].
      */
-    static long getEvent(long address, int i) {
-        return address + (SIZEOF_KQUEUEEVENT*i);
+    static MemorySegment getEvent(MemorySegment memoryHandle, int i) {
+        return kevent.asSlice(memoryHandle, i);
     }
 
     /**
      * Returns the file descriptor from a kevent (assuming it is in the ident field)
      */
-    static int getDescriptor(long address) {
-        return unsafe.getInt(address + OFFSET_IDENT);
+    static long getDescriptor(MemorySegment memoryHandle) {
+        return kevent.ident(memoryHandle);
     }
 
-    static short getFilter(long address) {
-        return unsafe.getShort(address + OFFSET_FILTER);
+    static short getFilter(MemorySegment memoryHandle) {
+        return kevent.filter(memoryHandle);
     }
 
-    static short getFlags(long address) {
-        return unsafe.getShort(address + OFFSET_FLAGS);
+    static short getFlags(MemorySegment memoryHandle) {
+        return kevent.flags(memoryHandle);
     }
 
     // -- Native methods --
 
-    private static native int keventSize();
-
-    private static native int identOffset();
-
-    private static native int filterOffset();
-
-    private static native int flagsOffset();
-
-    static native int create() throws IOException;
+    static public int register(int kqfd, int fd, int filter, int flags) {
+        int result;
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment keventMS = arena.allocate(kevent.layout());
+            kevent.ident(keventMS, fd);
+            kevent.filter(keventMS, (short) filter);
+            kevent.flags(keventMS, (short) flags);
+            // rest default to zero
+
+            // this do-while replaces restartable
```

Comment:
Future maintainers may ask what "replaces restartable" means, I think better to say that it calls kqueue if the syscall is interrupted (EINTR).